File: 147.对链表进行插入排序.py
#
# @lc app=leetcode.cn id=147 lang=python3
#
# [147] 对链表进行插入排序
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        # 我的
        root = ListNode(-float('inf'))
        root.next = head
        last = root
        while head:
            search = root
            nex = head.next
            if head.val >= last.val:
                last = last.next
            else:
                while search.next != head:
                    if head.val <= search.next.val:
                        tmp = search.next
                        search.next = head
                        head.next = tmp
                        last.next = nex
                        break
                    search = search.next
            head = nex

        return root.next

        # root = ListNode(-float('inf'))
        # while head:
        #     tmp = head.next
        #     pre = root
        #     while pre.next and pre.next.val < head.val:
        #         pre = pre.next
        #     head.next = pre.next
        #     pre.next = head
        #     head = tmp
        # return root.next
# @lc code=end

File: 147.对链表进行插入排序.java
/*
 * @lc app=leetcode.cn id=147 lang=java
 *
 * [147] 对链表进行插入排序
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode root = new ListNode(-Integer.MAX_VALUE);
        root.next = head;
        ListNode lastNode = root;
        ListNode searchNode;
        ListNode nextNode;
        ListNode tmp;
        while (head != null) {
            nextNode = head.next;
            if (head.val >= lastNode.val) {
                lastNode = lastNode.next;
            } else {
                searchNode = root;
                while (searchNode.next != head) {
                    if (head.val <= searchNode.next.val) {
                        tmp = searchNode.next;
                        searchNode.next = head;
                        head.next = tmp;
                        lastNode.next = nextNode;
                        break;
                    }
                    searchNode = searchNode.next;
                }
            }
            head = nextNode;
        }
        return root.next;
    }
}
// @lc code=end

